programs/bpf_loader/src/syscalls/cpi.rs

```diff
@@ -733,41 +739,89 @@ where
     Ok(accounts)
 }
 
+#[cfg(test)]
+static_assertions::const_assert_eq!(ACCOUNT_META_SIZE, 34);
+const ACCOUNT_META_SIZE: usize = size_of::<AccountMeta>();
+
 fn check_instruction_size(
     num_accounts: usize,
     data_len: usize,
     invoke_context: &mut InvokeContext,
 ) -> Result<(), EbpfError<BpfError>> {
-    let size = num_accounts
-        .saturating_mul(size_of::<AccountMeta>())
-        .saturating_add(data_len);
-    let max_size = invoke_context.get_compute_budget().max_cpi_instruction_size;
-    if size > max_size {
-        return Err(SyscallError::InstructionTooLarge(size, max_size).into());
+    if invoke_context
+        .feature_set
+        .is_active(&feature_set::loosen_cpi_size_restriction::ID)
+    {
+        let data_len = data_len as u64;
+        let max_data_len = MAX_CPI_INSTRUCTION_DATA_LEN;
+        if data_len > max_data_len {
+            return Err(SyscallError::MaxInstructionDataLenExceeded {
+                data_len,
+                max_data_len,
+            }
+            .into());
+        }
+
+        let num_accounts = num_accounts as u64;
+        let max_accounts = MAX_CPI_INSTRUCTION_ACCOUNTS as u64;
+        if num_accounts > max_accounts {
+            return Err(SyscallError::MaxInstructionAccountsExceeded {
+                num_accounts,
+                max_accounts,
+            }
+            .into());
+        }
+
+        invoke_context.get_compute_meter().consume(
```

Comment:
this is werid to have here, this function checks sizes, how about moving the consume up with the rest of the consumption stuff.  I'd be fine if you want to move it later during feature cleanup to simplify the featurization now